Dockerize

JerzySpendel opened the pull request.

grzegorz-wcislo commented on Dockerfile:
```diff
@@ -0,0 +1,13 @@
+FROM python:3.9
```
Maybe we can use the slim version instead

JerzySpendel replied:
Some packages need C compiler which is not there in slim version

grzegorz-wcislo commented on Dockerfile:
```diff
@@ -0,0 +1,13 @@
+FROM python:3.9
+
+RUN apt-get update -y
```
It's better to use something like:

```
RUN apt-get update -y \
  && apt-get upgrade -y \
  && apt get install -y --no-install-recommends postgresql-client \
  && rm -rf /var/lib/apt/lists/*
```

Using one RUN command results in fewer layers and a smaller image size

The rm at the end removes some not needed stuff that also reduces the image size

the `--no-install-recommends` prevents apt from installing more not needed stuff

grzegorz-wcislo commented on Dockerfile:
```diff
@@ -0,0 +1,13 @@
+RUN apt-get upgrade -y
+RUN apt-get install postgresql-client -y
+
+COPY . /code
```
We should also add a dockerignore file to prevent docker from copying not needed stuff from the repo (for example the `.git` folder or the nginx configuration files and whatnot.

grzegorz-wcislo commented on Dockerfile:
```diff
@@ -0,0 +1,13 @@
+
+RUN pip install poetry
+RUN poetry config virtualenvs.create false
+RUN poetry install --no-root
```
We should probably also add a user and not run everything as root, I think this is an issue for later though

JerzySpendel replied:
So far I have been assuming that it's fine to do everything as a root in docker containers and I think this is accepted convention to do so? :D

This `--no-root` option here is not what it might look like, it's just as parameter that prevents installing this whole application as a package to current venv. If I didn't pass this `--no-root` option then poetry would install whole `physionet` as a package and then I could do `import physionet` in current venv (which would be the global python interpreter in that case). Poetry is often used as a packaging framework if somebody write a library/framework or something, in these cases it's useful.

grzegorz-wcislo replied:
There are some security concerns when running as root:

https://security.stackexchange.com/questions/176206/docker-runs-container-processes-as-root-should-i-be-worried
https://americanexpress.io/do-not-run-dockerized-applications-as-root/

grzegorz-wcislo commented on physionet-django/project/admin.py:
```diff
@@ -3,7 +3,7 @@
 from django.db.models import CharField, TextField
 from django.forms import TextInput, Textarea
-from background_task.models_completed import CompletedTask
+#from background_task.models_completed import CompletedTask
```
Why did you remove it?

JerzySpendel replied:
Ok, uncommented again. I had too new version which didn't have such module `models_completed` and I was wondering why it's not working for me. So I figured out in `django-background-tasks` 1.2.0 there IS `models_completed` module and in 1.2.6 there IS NOT which is a breaking change in a x.y.Z version change, not the best thing to do.

grzegorz-wcislo commented on wait-for-it.sh:
```diff
@@ -0,0 +1,182 @@
+#!/usr/bin/env bash
```
Very nice addition, i really like this script 😛

JerzySpendel replied:
:D

grzegorz-wcislo commented on docker-compose.yml:
```diff
@@ -0,0 +1,37 @@
+version: "3.9"
```
This docker-compose is intended only for local development, right?

I'd recommend adding 2 more volumes (named volumes, not the directory ones), one for media (/code/media) and one for the database.

JerzySpendel replied:
There's already volume for /code, so why another for /code/media?

grzegorz-wcislo reviewed (changes requested):
Nice work :)

I request changes because of the Dockerfile and dockerignore stuff. Also please remember to add trailing whitespaces to all the files (github shows a missing one with the red circle at the end of the file)

rafalcymerys commented on physionet-django/physionet/settings/settings.py:
```diff
@@ -1,45 +1,47 @@
-DEFAULT_FROM_EMAIL = 'PhysioNet Automated System <noreply@physionet.org>'
-CONTACT_EMAIL = 'PhysioNet Contact <contact@physionet.org>'
-SERVER_EMAIL = 'PhysioNet System <root@physionet.org>'
+DEFAULT_FROM_EMAIL = config("DEFAULT_FROM_EMAIL")
```
It would be great to use old values as default ones, not to break compatibility for physionet's existing deployments

rafalcymerys commented on physionet-django/physionet/settings/settings.py:
```diff
@@ -1,45 +1,47 @@
-EMAIL_HOST = 'localhost'
+EMAIL_BACKEND = 'django.core.mail.backends.console.EmailBackend' if DEBUG else 'django.core.mail.backends.smtp.EmailBackend'
+EMAIL_HOST = config("EMAIL_HOST")
 EMAIL_PORT = 25
```
Let's also parameterize SMTP port etc. - if we want to use AWS SES, we'll definitely need a different port + authentication information

grzegorz-wcislo committed:
Remove unused imports

grzegorz-wcislo committed:
Add poetry configuration

Poetry is a python dependency management tool.
Using poetry makes managing dependency versions easier than
using only `requirements.txt`.

`poetry.lock` is an automatically generated lockfile that pins
dependency versions.

grzegorz-wcislo committed:
Change query separator in tests

Changes query arguments separators from `;` to `&` in lightwave tests.
This matches the configuration that is used in production.

grzegorz-wcislo committed:
Extract configuration to environment variables

Unifies the configuration by merging development, staging and production
configuration files.

The defaults are taken from previous default production settings.

This will require changes to the `.env` files used in staging and
production.

grzegorz-wcislo committed:
Add Dockerfile and docker-compose

Adds a Dockerfile and docker-compose.yml file for local development and
testing.

Adds scripts in the `docker` directory:
  - dev-entrypoint.sh : Used by docker-compose to create required
    directories and run migrations.
  - init.sh : Used by docker-compose to create the development and test
    databases.
  - wait-for-it.sh : Used by dev-entrypoint.sh to wait for the database
    to be available.

Updates the `resetdb.py` script to handle remote databases in both sudo and
non-sudo environments.

Updates README.md to reflect the changes.

grzegorz-wcislo committed:
Use psycopg2 instead of psycopg2-binary

Building the package from source makes the Dockerfile more portable.

Authors of psycopg2 recommend installing from source.
